### PR TITLE
Perspective correction

### PR DESCRIPTION
**Pull request by campagnola**
This fixes perspective transforms. Summary:

* OpenGL uses homogeneous coordinates; our transforms need to do the same.
* Corrected STTransform to work correctly when w != 1
* PerspectiveTransform is now just a thin wrapper around AffineTransform. Might just combine these both into MatrixTransform in a future PR.
* Fixed ImageVisual to work correctly in impostor mode when its transforms affect z
* Added a script that I used to test perspective. This will be converted into a unit test with image comparison in a future PR.
* There is a bizarre bug on my machine that I suspect is an OpenGL bug. The console example gives me a warning:

```
INFO: Variable u_scale is not an active uniform
```

This is due to the STTransform map function, which now looks like:

```
        vec4 st_transform_map(vec4 pos) {
            return vec4(pos.xyz * $scale.xyz + $translate.xyz * pos.w, pos.w);
        }
```

If I switch back to the old implementation, the warning goes away:

```
        vec4 st_transform_map(vec4 pos) {
            return (pos * $scale) + $translate * pos.w;
        }
```

I can't see why the $scale uniform would be optimized out of the shader in either case. I'm betting Travis won't have the same error, though..

### CODE REVIEW

File: vispy/scene/cameras/tests/perspective_test.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        image.transform = scene.STTransform(translate=(-12.8, -12.8),
+                                            scale=(0.1, 0.1))
+        v.add(image)
+        images.append(image)
+        
+
+@canvas.connect
+def on_key_press(ev):
+    if ev.key.name == '1':
+        print "Image method: impostor"
```
Comment (Eric89GXL) [inline vispy/scene/cameras/tests/perspective_test.py]: all of these print statements are non-Py3k compatible

File: vispy/scene/cameras/tests/perspective_test.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        v.camera.distance = 30
+        #v.camera = 'panzoom'
+        #v.camera.aspect = 1
+        
+        views.append(v)
+        image = scene.visuals.Image(imdata, method='impostor', grid=(4, 4))
+        image.transform = scene.STTransform(translate=(-12.8, -12.8),
+                                            scale=(0.1, 0.1))
+        v.add(image)
+        images.append(image)
```
Comment (Eric89GXL) [inline vispy/scene/cameras/tests/perspective_test.py]: Maybe a stupid question, but why have four of these?

File: vispy/scene/cameras/tests/perspective_test.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        for vb in views:
+            vb.clip_method = 'fragment'
+    elif ev.key.name == '4':
+        print "Viewbox method: viewport"
+        for vb in views:
+            vb.clip_method = 'viewport'
+    elif ev.key.name == '5':
+        print "Viewbox method: fbo"
+        for vb in views:
+            vb.clip_method = 'fbo'
```
Comment (Eric89GXL) [inline vispy/scene/cameras/tests/perspective_test.py]: Should add `.update()` calls after each of these, right? Or maybe just one at the end (updating is okay even if a non-meaningful key was pressed).

Comment (Eric89GXL) [inline vispy/scene/cameras/tests/perspective_test.py]: ...or does it auto-update because of the changed param?

Comment (Eric89GXL) [inline vispy/scene/cameras/tests/perspective_test.py]: FYI 1/2 seems to look the same, 3/4 look the same and 5 (FBO clipping) is messed up on my system. Is this what you get?

File: vispy/scene/cameras/tests/perspective_test.py (hunk 2 above)
Comment (campagnola) [inline vispy/scene/cameras/tests/perspective_test.py]: Some of the bugs I was dealing with were dependent on the position of the viewbox within the canvas.

**Comment by Eric89GXL** [comment]
You mean you have that warning in `examples/basics/scene/console.py`? I do not on OSX (NVIDIA mobile GPU), I can test Linux tomorrow.

File: vispy/scene/cameras/tests/perspective_test.py (hunk 3 above)
Comment (campagnola) [inline vispy/scene/cameras/tests/perspective_test.py]: Yes, setting vb.clip_method triggers an update.
Yes, that is the output I get. The FBO issue is for another day.

**Comment by Eric89GXL** [comment]
Looks like Travis is angry because `py.test` tried to run your `perspective_test.py` as a test. Renaming the file (not to have `test` in the name) should fix it, I think.

**Comment by campagnola** [comment]
Ah! Travis has the same error. It does make sense, though. The shader looks like:
```
uniform vec4 u_scale;
uniform vec4 u_translate;
vec4 st_transform_map(vec4 pos) {
    return vec4(pos.xyz * u_scale.xyz + u_translate.xyz * pos.w, pos.w);
}

attribute vec2 a_position;
. . .
void main (void) {
    vec4 pos = st_transform_map(vec4(0., 0., 0., 1.));
    gl_Position = pos + vec4(a_position * u_logical_scale, 0., 0.);
    . . . 
}

```
Since the input to `st_transform_map` has xyz=0, `u_scale` can be optimized out. Will post a fix soon.

**Comment by campagnola** [comment]
Green light! @almarklein would you like a peek?

**Comment by almarklein** [comment]
Changes look good to me. However, for volume rendering this new perspective does not work as it should.

**Comment by campagnola** [comment]
Check out the last commit--this fixes perspective in volume and also no longer requires face subdivision. The downside is that it moves the ray calculation to the fragment shader, which may slow the rendering down (but its probably not significant compared to the for-loop).

There is an issue with clipping that I wasn't able to fix. I don't quite have a handle on how the shaders work. Do you want to work on this in a sub-PR?

File: vispy/visuals/volume.py [hunk 4]
```diff
@@ -51,37 +51,15 @@
...
-    
-    // Calculate a clip plane (in texture coordinates) for the camera
-    // position, in case that the camera is inside the volume.
-    pos_in_cam1.zw = vec2(u_cameraclip, 1.0);
-    vec3 cameraposinvol = $viewtransformi(pos_in_cam1).xyz;
-    cameraposinvol /= u_shape;  // express in texture coords
-    v_clipplane.xyz = v_ray;
-    v_clipplane.w = dot(v_clipplane.xyz, cameraposinvol);
+    v_position = a_position;
+    gl_Position = $transform(vec4(v_position, 1));
```
Comment (almarklein) [inline vispy/visuals/volume.py]: 1.0

**Comment by almarklein** [comment]
Wow, I truly believed it was not possible to get correct perspective without partitioning. You blew my mind :)

As for clipping, I suspect its related to the `w` factor which is now different. Also note the artifacts when fov=0.

File: vispy/visuals/volume.py (hunk 4 above)
Comment (campagnola) [inline vispy/visuals/volume.py]: Is a picky requirement of some GL drivers?

Comment (almarklein) [inline vispy/visuals/volume.py]: Yes, we've had problems with some drivers because of this. Also, it will *always* cause compilation errors on WebGL.

**Comment by Eric89GXL** [comment]
@campagnola looks like you addressed @almarklein's comments, except for clipping / fov=0 artifacts? Or are those fixed now / in a subsequent PR?

**Comment by campagnola** [comment]
Last commit fixes clipping and moves the transformations back to the vertex shader where they belong.

**Comment by Eric89GXL** [comment]
Good to go from your end, then @campagnola?

**Comment by campagnola** [comment]
Yep. There is still an edge artifact, but that is present in master as well.

**Comment by Eric89GXL** [comment]
Looks like Travis is failing on GLFW, almost certainly unrelated to your changes. Looks like we've probably found another bug for them :(

**Comment by almarklein** [comment]
> but that is present in master as well.

mmm, you're right. I hadn't seen it yet (except for before I fixed it early on. I guess I did not fix it enough)

File: vispy/visuals/volume.py [hunk 5]
```diff
@@ -51,37 +50,29 @@
...
 attribute vec3 a_texcoord;
 uniform vec3 u_shape;
-uniform float u_cameraclip;
 
 varying vec3 v_texcoord;
-varying vec3 v_ray;
-varying vec4 v_clipplane;
+varying vec3 v_position;
+varying vec4 v_position2;
+varying vec4 clipplane;
```
Comment (almarklein) [inline vispy/visuals/volume.py]: Not used

File: vispy/visuals/volume.py [hunk 6]
```diff
@@ -105,39 +96,44 @@
...
-    // Discart front facing
-    //if (!gl_FrontFacing)
-    //    discard;
-    
-    // Uncomment this to show a grid of the backfaces
-    //vec3 pcd = v_texcoord * u_shape;
-    //for (int d=0; d<3; d++)
-    //    if (pcd[d] > 1 && pcd[d] < (u_shape[d]-1) && sin(pcd[d]) > 0.9)
-    //            discard;
+    // Calculate ray. In the fragment shader we do another normalization
```
Comment (almarklein) [inline vispy/visuals/volume.py]: This comment needs an update

File: vispy/visuals/volume.py [hunk 7]
```diff
@@ -628,7 +614,7 @@ def _create_vertex_data(self):
         
         # Get indices and vertices for triangles. 
         indices, tex_coord, ver_coord = self._calc_coords(tex_coord, ver_coord,
-                                                          div)
+                                                          2)
```
Comment (almarklein) [inline vispy/visuals/volume.py]: I suppose you can remove lots of lines above in this function. Which is awesome, it should even increase performance a wee bit.

File: vispy/visuals/volume.py [hunk 8]
```diff
@@ -8,14 +8,13 @@
 
 In Python, we define the six faces of a cuboid to draw, as well as
 texture cooridnates corresponding with the vertices of the cuboid. In
-the vertex shader, we establish the ray direction and pass it to the
-fragment shader as a varying. In the fragment shader, we have the
-texture coordinate as the starting point of the ray, and we know the
-ray direction. We then calculate the number of steps and use that number
+the vertex shader, we compute the near clipping position along the view
+ray and pass it to the fragment shader as a varying. In the fragment shader, 
+we compute the ray direction and calculate the number of steps to use
```
Comment (almarklein) [inline vispy/visuals/volume.py]: It seems this needs an update too.

**Comment by almarklein** [comment]
Apart for some small comments, this LGTM. Nice work!

**Comment by campagnola** [comment]
> mmm, you're right. I hadn't seen it yet (except for before I fixed it early on. I guess I did not fix it enough)

I think it is caused by the emulated 3D texture interpolating between z-slices. I have another PR coming up that touches some of that code, though, so let's hold off fixing it until then.

**Comment by Eric89GXL** [comment]
@campagnola is this good to go from your end, then?

**Comment by Eric89GXL** [comment]
(I restarted your test now that GLFW is fixed and the CIs are happy)

**Comment by campagnola** [comment]
Cool, I'm happy with this one.

**Comment by Eric89GXL** [comment]
Great, merged.

**Comment by campagnola** [comment]
Huzzah!

**Pull request merged**